zipline-loader/src/commonTest/kotlin/app/cash/zipline/loader/internal/cache/ZiplineCacheFaultsTest.kt

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright (C) 2024 Cash App
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package app.cash.zipline.loader.internal.cache
+
+import assertk.assertThat
+import assertk.assertions.each
+import assertk.assertions.isEqualTo
+import assertk.assertions.isIn
+import assertk.assertions.isLessThanOrEqualTo
+import kotlin.test.Test
+import kotlinx.coroutines.runBlocking
+
+/**
+ * This is almost like a fuzz test: in each method it runs many iterations of the same operations,
+ * but each iteration injects a failure after a different number of writes.
+ *
+ * Because the behavior or the file system is different on each run, the test is lenient in what it
+ * expects from the cache: it accepts a wide variety of download counts. But regardless of whether
+ * the file system accepts writes, the cache should never return invalid results.
+ */
+class ZiplineCacheFaultsTest {
+  @Test
+  fun openMissHitClose(): Unit = runBlocking {
+    // How many writes are attempted in this test in the happy path. Determined experimentally!
+    val noFailuresWriteCount = 20
+
+    for (i in 0 until noFailuresWriteCount + 1) {
```

Comment:
until is going to do `- 1`
```suggestion
    for (i in 0..noFailuresWriteCount) {
```

Reply:
Got 'em in the follow-up.